replica/table_sstable_set.cc

```diff
@@ -0,0 +1,240 @@
+            auto* sg = _table.storage_group_for_token(pos.token());
+            _cur_set = sg->make_compound_sstable_set();
+            _cur_selector.emplace(_cur_set->make_incremental_selector());
+            auto next_token = sg->main_compaction_group()->next_compaction_group_token();
+            _lowest_next_position = next_token.is_maximum()
+                    ? _lowest_next_position = dht::ring_position_ext::max()
+                    : dht::ring_position_ext::starting_at(next_token);
+        }
+
+        auto res = _cur_selector->select(pos);
+        // Return all sstables selected on the requested position from the first matching sstable set.
+        // This assumes that the underlying sstable sets are disjoint in their token ranges so
+        // only one of them contain any given token.
+        auto sstables = std::move(res.sstables);
+        // Return the lowest next position, such that this function will be called again to select the
+        // lowest next position from the selector which previously returned it.
+        // Until the current selector is exhausted. In that case,
+        // jump to the next compaction_group sstable set.
+        dht::ring_position_ext lowest_next_position = res.next_position;
+        if (lowest_next_position.is_max()) {
+            _cur_set = {};
+            _cur_selector.reset();
+            std::swap(lowest_next_position, _lowest_next_position);
+        }
+
+        tlogger.trace("table_incremental_selector: select pos={}: returning {} sstables, next_pos={}", pos, sstables.size(), lowest_next_position);
+        return std::make_tuple(std::move(current_range), std::move(sstables), std::move(lowest_next_position));
+    }
+};
+
+sstables::sstable_set_impl::selector_and_schema_t table_sstable_set::make_incremental_selector() const {
+    return std::make_tuple(std::make_unique<table_incremental_selector>(_table), *_table.schema());
+}
+
+flat_mutation_reader_v2
+table_sstable_set::create_single_key_sstable_reader(
+        replica::column_family* cf,
+        schema_ptr schema,
+        reader_permit permit,
+        utils::estimated_histogram& sstable_histogram,
+        const dht::partition_range& pr,
+        const query::partition_slice& slice,
+        tracing::trace_state_ptr trace_state,
+        streamed_mutation::forwarding fwd,
+        mutation_reader::forwarding fwd_mr,
+        const sstable_predicate& predicate) const {
+    // The singular partition_range start bound must be engaged.
+    auto token = pr.start()->value().token();
+    auto* sg = _table.storage_group_for_token(token);
+    auto set = sg->make_compound_sstable_set();
+    return set->create_single_key_sstable_reader(cf, std::move(schema), std::move(permit), sstable_histogram, pr, slice, trace_state, fwd, fwd_mr, predicate);
+}
+
+lw_shared_ptr<sstables::sstable_set> table::make_table_sstable_set() {
+    return make_lw_shared(table_sstable_set::make(*this));
+}
+
+} // namespace replica
```

Comment:
tests?

Reply:
Pretty much everything that reads from a table or uses the table's incremental selector uses this new code, so it's pretty well covered.  I can also try to integrate the table incremental selector in the respective unit test but it's not trivial.

Reply:
@bhalevy I suggest you look at this:

https://github.com/scylladb/scylladb/pull/13666/commits/46b1430d2f1b826e86d2e25bc724358e9255fa5c
https://github.com/scylladb/scylladb/pull/13666/commits/6b267744c4ef2b8fa6751b17a79444a8a9462ab9